File: app/lib/services/road_graph_engine/route_search.dart
import 'package:collection/collection.dart';
import '../../models/road_segment.dart';
import '../../models/route_result.dart';
import 'graph.dart';

/// prototype/src/routeSearch.js のDart移植（Dijkstra）。
/// コスト = 距離 × (1 - shadeWeight × shadowScore)。技術検証プロトタイプで
/// 200回反復・平均0.1msの実用速度を確認済み（RESULTS.md参照、小規模データでの目安）。
RouteResult? searchRoute({
  required RoadGraph graph,
  required String originNodeId,
  required String destNodeId,
  double shadeWeight = 0.6,
}) {
  if (!graph.nodeById.containsKey(originNodeId) || !graph.nodeById.containsKey(destNodeId)) {
    return null;
  }

  final dist = <String, double>{originNodeId: 0};
  final prevEdge = <String, String>{};
  final prevNode = <String, String>{};
  final visited = <String>{};

  final queue = HeapPriorityQueue<_Entry>((a, b) => a.priority.compareTo(b.priority));
  queue.add(_Entry(originNodeId, 0));

  while (queue.isNotEmpty) {
    final current = queue.removeFirst();
    if (visited.contains(current.nodeId)) continue;
    visited.add(current.nodeId);
    if (current.nodeId == destNodeId) break;

    for (final edge in graph.adjacency[current.nodeId] ?? const <GraphEdge>[]) {
      final neighbor = edge.otherEnd(current.nodeId);
      if (visited.contains(neighbor)) continue;
      final edgeCost = edge.distanceM * (1 - shadeWeight * edge.shadowScore);
      final newCost = current.priority + edgeCost;
      if (newCost < (dist[neighbor] ?? double.infinity)) {
        dist[neighbor] = newCost;
        prevEdge[neighbor] = edge.id;
        prevNode[neighbor] = current.nodeId;
        queue.add(_Entry(neighbor, newCost));
      }
    }
  }

  if (!dist.containsKey(destNodeId)) return null;

  final pathNodeIds = <String>[destNodeId];
  final segments = <RoadSegment>[];
  var cur = destNodeId;
  var totalDistanceM = 0.0;
  var totalComfortWeighted = 0.0;

  while (cur != originNodeId) {
    final edgeId = prevEdge[cur];
    final prev = prevNode[cur];
    if (edgeId == null || prev == null) break;
    final edge = graph.edgeById[edgeId]!;
    final fromNode = graph.nodeById[edge.from]!;
    final toNode = graph.nodeById[edge.to]!;
    final segment = RoadSegment(
      id: edge.id,
      from: fromNode,
      to: toNode,
      distanceM: edge.distanceM,
      baseShadowScore: edge.shadowScore,
      aggregatedShadeScore: edge.shadowScore,
    );
    segments.add(segment);
    totalDistanceM += edge.distanceM;
    totalComfortWeighted += segment.comfortScore * edge.distanceM;
    pathNodeIds.add(prev);
    cur = prev;
  }

  final orderedNodeIds = pathNodeIds.reversed.toList();
  final nodes = orderedNodeIds.map((id) => graph.nodeById[id]!).toList();
  final averageComfort = totalDistanceM > 0 ? totalComfortWeighted / totalDistanceM : 0.0;

  return RouteResult(
    nodes: nodes,
    segments: segments.reversed.toList(),
    distanceM: totalDistanceM,
    averageComfortScore: averageComfort,
  );
}

class _Entry {
  _Entry(this.nodeId, this.priority);
  final String nodeId;
  final double priority;
}

File: functions/src/routeSearch.js
// prototype/src/routeSearch.js と同一。
// 重み付き最短経路探索（Dijkstra）。
// コスト = 距離 × (1 - shadeWeight × shadowScore) — 影が濃い区間ほどコストが下がり優先されやすくなる。
import MinHeap from './minHeap.js';

/**
 * @param {ReturnType<import('./buildGraph.js').buildGraph>} graph
 * @param {Map<string, number>} shadowScores edgeId -> 0..1
 * @param {string} originNodeId
 * @param {string} destNodeId
 * @param {{shadeWeight?: number}} [weightPrefs]
 * @returns {{path: string[], distanceM: number, cost: number} | null}
 */
export function searchRoute(graph, shadowScores, originNodeId, destNodeId, weightPrefs = {}) {
  const shadeWeight = weightPrefs.shadeWeight ?? 0.5; // 0=距離最優先, 1=安心スコア最優先

  if (!graph.nodeById.has(originNodeId) || !graph.nodeById.has(destNodeId)) return null;

  const edgeById = graph.edgeById ?? new Map(graph.edges.map((e) => [e.id, e]));

  const dist = new Map([[originNodeId, 0]]);
  const prevEdge = new Map();
  const prevNode = new Map();
  const visited = new Set();
  const heap = new MinHeap();
  heap.push(originNodeId, 0);

  while (!heap.isEmpty()) {
    const { item: current, priority: currentCost } = heap.pop();
    if (visited.has(current)) continue;
    visited.add(current);
    if (current === destNodeId) break;

    const neighbors = graph.adjacency.get(current) ?? [];
    for (const { to, edgeId, distanceM } of neighbors) {
      if (visited.has(to)) continue;
      const shadowScore = shadowScores.get(edgeId) ?? 0;
      const edgeCost = distanceM * (1 - shadeWeight * shadowScore);
      const newCost = currentCost + edgeCost;
      if (newCost < (dist.get(to) ?? Infinity)) {
        dist.set(to, newCost);
        prevEdge.set(to, edgeId);
        prevNode.set(to, current);
        heap.push(to, newCost);
      }
    }
  }

  if (!dist.has(destNodeId)) return null;

  // 経路復元
  const path = [destNodeId];
  let cur = destNodeId;
  let distanceM = 0;
  while (cur !== originNodeId) {
    const edgeId = prevEdge.get(cur);
    const prev = prevNode.get(cur);
    if (!edgeId || !prev) break;
    const edge = edgeById.get(edgeId);
    distanceM += edge?.distanceM ?? 0;
    path.push(prev);
    cur = prev;
  }
  path.reverse();

  return { path, distanceM, cost: dist.get(destNodeId) };
}
